src/service/position-observer-impl.js

```diff
@@ -0,0 +1,389 @@
+   * @param {function(PositionInViewportEntryDef)} handler
+   * @return {?Object}
+   */
+  observe(element, fidelity, handler) {
+    // If the element already exists in current observeEntries, do nothing
+    for (let i = 0; i < this.entries_.length; i++) {
+      if (this.entries_[i].element === element) {
+        dev().error(TAG, 'should observe same element only once');
+        return null;
+      }
+    }
+
+    // make entry into a class
+    const entry = {
+      element,
+      handler,
+      fidelity,
+      position: null,
+      turn: (fidelity == PositionObserverFidelity.LOW) ?
+          Math.floor(Math.random() * LOW_FEDELITY_FRAME_COUNT) : 0,
+      trigger: function(position) {
+        const prePos = entry.position;
+        if (prePos
+            && layoutRectEquals(prePos.positionRect, position.positionRect)
+            && layoutRectEquals(prePos.viewportRect, position.viewportRect)) {
+          // position doesn't change, do nothing.
+          return;
+        }
+        if (layoutRectsOverlap(position.positionRect, position.viewportRect)) {
+          entry.position = position;
+          // Only call handler if entry element overlap with viewport.
+          try {
+            entry.handler(position);
+          } catch(err) {}
+        } else if (entry.position) {
+          // Need to notify that element gets outside viewport
+          // NOTE: This is required for inabox position observer.
+          entry.position = null;
+          position.positionRect = null;
+          entry.handler(position);
+        }
+      },
+    };
+
+    this.entries_.push(entry);
+
+    if (this.entries_.length == 1) {
+      this.startCallback();
+    }
+    this.updateEntryPosition_(entry);
+    return entry;
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {PositionObserverFidelity} fidelity
+   */
+  changeFidelity(element, fidelity) {
+    for (let i = 0; i < this.entries_.length; i++) {
+      const entry = this.entries_[i];
+      if (entry.element === element) {
+        entry.fidelity = fidelity;
+        if (fidelity == PositionObserverFidelity.HIGH) {
+          entry.turn = 0;
+        }
+        return;
+      }
+    }
+    dev().error(TAG, 'cannot change fidelity on unobserved element');
+  }
+
+  /**
+   * @param {!Element} element
+   */
+  unobserve(element) {
+    for (let i = 0; i < this.entries_.length; i++) {
+      if (this.entries_[i].element === element) {
+        this.entries_[i].handler = null;
+        this.entries_.splice(i, 1);
+        if (this.entries_.length == 0) {
+          this.stopCallback();
+        }
+        return;
+      }
+    }
+    dev().error(TAG, 'cannot unobserve unobserved element');
+  }
+
+  /** @param {!Object} unusedEntry */
+  updateEntryPosition_(unusedEntry) {}
+
+  startCallback() {}
+
+  stopCallback() {}
+
+  increaseFidelityCallback() {}
+
+  decreaseFidelityCallback() {}
+}
+
+export class AmpPagePositionObserver extends AbstractPositionObserver {
```

Comment:
What is "AmpPage"? Is it "ampdoc"? "AMP Host"? I think these terms are more specific and less ambiguous.